Add features to bulk import for assessing and tracking results

graykr opened the pull request:
I've added a few features to bulk_import.py to help with assessing and tracking the results of the bulk import jobs:
* generate a csv and txt report for each bulk import to log and track data for each file in the batch
* provide an option to download the csv output of the import jobs from ASpace (provides more detail on lines in the csv with possible import issues)
* if downloaded, parse the output files and provide a count of the issues in the report document so we can easily identify which ones require additional review
* check whether the resource already has children before importing a csv finding aid (to prevent accidentally adding to an existing finding aid)

graykr committed:
feat: save txt and csv reports for tracking bulk imports

graykr committed:
feat: add option to download output files for bulk import

graykr committed:
docs: add descriptions for new functions created

graykr committed:
fix: results for bulk import should also include files with errors

graykr committed:
feat: check for whether there are already children of a resource before trying to import a csv for it

graykr committed:
feat: parse the csv output generated by the import job if downloaded and log number of issues noted therein

graykr committed:
fix: check whether the import type is "ao" before checking if the resource already has children

also remove extraneous print statement

leonelramirez commented on arcflow/utils/bulk_import.py:
```diff
@@ -128,8 +172,122 @@ def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false'):
+def check_job_status(client, repo_id, job_id):
+    """Function to check whether a job has completed (and thus output files are ready)."""
+    while True:
+        job_status_response = client.get(f'/repositories/{repo_id}/jobs/{job_id}')
```
For parameter naming consistency.
```suggestion
def check_job_status(asnake_client, repo_id, job_id):
    """Function to check whether a job has completed (and thus output files are ready)."""
    while True:
        job_status_response = asnake_client.get(f'/repositories/{repo_id}/jobs/{job_id}')
```

leonelramirez commented on arcflow/utils/bulk_import.py:
```diff
@@ -146,13 +304,21 @@ def main():
-
+    
+    if args.save_output_files:
+        retrieve_job_output(args.dir, import_report)
```
Since `retrieve_job_output` modifies the output of `csv_bulk_import`, I recommend to put `retrieve_job_output` inside `csv_bulk_import` to maintain logical cohesion, and add an `asnake_client` parameter to `retrieve_job_output` (this also eliminates the need to instantiate an additional `ASnakeClient`).

Part 1 of 4
```suggestion
    import_report = csv_bulk_import(
            csv_directory=args.dir,
            load_type=args.load_type,
            only_validate='true' if args.only_validate else 'false',
            save_output_files=args.save_output_files)
```

leonelramirez commented on arcflow/utils/bulk_import.py:
```diff
@@ -64,6 +67,22 @@ def get_resource_id_from_ead(ead_id, asnake_client):
+    report_dict["error"] = error_string
+    print(error_string)
 
 def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false'):
```
Since `retrieve_job_output` modifies the output of `csv_bulk_import`, I recommend to put `retrieve_job_output` inside `csv_bulk_import` to maintain logical cohesion, and add an `asnake_client` parameter to `retrieve_job_output` (this also eliminates the need to instantiate an additional `ASnakeClient`).

Part 2 of 4
```suggestion
def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false', save_output_files=False):
```

leonelramirez commented on arcflow/utils/bulk_import.py:
```diff
@@ -128,8 +172,122 @@ def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false'):
+
+        bulk_import_report.append(file_import_report)
         print(json.dumps(import_job, indent=4))
+    return bulk_import_report
```
Since `retrieve_job_output` modifies the output of `csv_bulk_import`, I recommend to put `retrieve_job_output` inside `csv_bulk_import` to maintain logical cohesion, and add an `asnake_client` parameter to `retrieve_job_output` (this also eliminates the need to instantiate an additional `ASnakeClient`).

Part 3 of 4
```suggestion
        bulk_import_report.append(file_import_report)                
        print(json.dumps(import_job, indent=4))
        
    if save_output_files:
        retrieve_job_output(csv_directory, bulk_import_report, client)
        
    return bulk_import_report
```

leonelramirez commented on arcflow/utils/bulk_import.py:
```diff
@@ -128,8 +172,122 @@ def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false'):
+            continue
+        job_files = client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files").json()
+        output_file_id = max(job_files)
+        response = client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files/{output_file_id}")
```
Since `retrieve_job_output` modifies the output of `csv_bulk_import`, I recommend to put `retrieve_job_output` inside `csv_bulk_import` to maintain logical cohesion, and add an `asnake_client` parameter to `retrieve_job_output` (this also eliminates the need to instantiate an additional `ASnakeClient`).

Part 4 of 4
```suggestion
def retrieve_job_output(path, report_list, asnake_client):
    """Function to retrieve and save last created output files for each job in the bulk import."""
    for row in report_list:
        if "results_id" not in row:
            continue
        repo_id = row["repo_id"]
        job_id = row["results_id"]
        identifier = row["identifier"]
        if not check_job_status(client, repo_id, job_id):
            print(f"Error downloading files for job {job_id}")
            continue
        job_files = asnake_client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files").json()
        output_file_id = max(job_files)
        response = asnake_client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files/{output_file_id}")
```

graykr committed:
refactor: apply changes from code review

call retrieve_job_output from inside csv_bulk_import instead of separately; rename client parameter name for consistency
